core/model/listeners.py

```diff
@@ -18,6 +24,128 @@
 from .licensing import LicensePool
 from .work import Work, add_work_to_customlists_for_collection
 
+# TODO: Remove this when we drop support for Python 3.9
+if sys.version_info >= (3, 10):
+    from typing import ParamSpec
+else:
+    from typing_extensions import ParamSpec
+
+
+class State(Enum):
+    NEW = "new"
+    DELETED = "deleted"
+    DIRTY = "dirty"
+    ANY = "any"
+
+
+_before_flush_hooks: Dict[
+    Tuple[Type[Base], State], List[Callable[[Session, Base], None]]
+] = defaultdict(list)
+_before_flush_trigger_hooks: Dict[
+    Callable[[Session], None], Set[Type[Base]]
+] = defaultdict(set)
+
+
+P = ParamSpec("P")
+
+
+def before_flush(
+    model: Type[Base], state: State
+) -> Callable[[Callable[P, None]], Callable[P, None]]:
+    """
+    Decorator to register a function to be called before a flush.
+
+    The decorated function will be called when a model of the given type is in the given state. The function
+    will be called with two arguments: the session and the instance of the model that triggered the flush.
+    """
+
+    def decorator(func: Callable[P, None]) -> Callable[P, None]:
+        _before_flush_hooks[(model, state)].append(func)  # type: ignore[arg-type]
+        return func
+
+    return decorator
+
+
+def before_flush_trigger(
```

Comment:
How is this different from `before_flush(<model>, ANY)` apart from the multiple models parameter?

Reply:
Looks like a lot of special code for the same use case

Reply:
It only fires the listener once, if any of the models are changed. Rather then firing it repeatedly for each instance that has changed. This is used for `site_configuration_has_changed`. If we don't have this triggered style decorator, if a flush has updated 100 ConfigurationSettings, we will call `site_configuration_has_changed` for each of these.

Reply:
Since `site_configuration_has_changed` acquires a lock, and emits SQL directly, this ends up being a lot of unnecessary work. I don't see exactly how to support this without an additional decorator here. Honestly I think implementing decorators might be more work then its worth here, its taken almost as much time as it did to do the upgrade.

Would it be better to just drop the commit that adds decorators here, and go back to the previous implementation.

Reply:
I don't know why but this is much easier to read. Thanks for doing that :+1:

Reply:
Thanks for the review and feedback! It made this a lot better. 

I rushed the first implementation of decorators and looking back at it, it wasn't very readable. Think it's in a much better place now.